### PR TITLE
(SERVER-1922) Set puppet log level from logback

### PR DESCRIPTION
**Pull request by adrienthebo**
In the early days of Puppetserver, the Puppet log level was maxed out in
order to pass all filtering control to logback. This approach didn't
take into account the potential cost of generating log messages and in
the real world can have performance implications.

This pull request resolves this problem by using the logback
puppetserver log level to set the Puppet log level. Because there is not
a 1:1 mapping of logback levels to Puppet levels the notice log level
cannot be set; if notice level information is needed then the info level
must be set.

### CODE REVIEW

File: src/ruby/puppetserver-lib/puppet/server/puppet_config.rb [hunk 1]
```diff
@@ -29,9 +29,6 @@ def self.initialize_puppet(puppet_config)
       Puppet[:always_retry_plugins] = false
     end
 
-    # Crank Puppet's log level all the way up and just control it via logback.
```
Comment (Magisus) [inline src/ruby/puppetserver-lib/puppet/server/puppet_config.rb]: Is there code to actually do this controlling that should be ripped out? Or was it sort of automatic?

File: src/ruby/puppetserver-lib/puppet/server/logger.rb [hunk 2]
```diff
@@ -27,5 +27,25 @@ def handle(msg)
 class Puppet::Server::Logger
   def self.init_logging
     Puppet::Util::Log.newdestination(:logback)
+    Puppet[:log_level] = level_from_logback(get_logger)
+  end
+
+  def self.level_from_logback(logger)
+    case
+    when logger.isDebugEnabled()
+      'debug'
```
Comment (MosesMendoza) [inline src/ruby/puppetserver-lib/puppet/server/logger.rb]: should these be symbols?

Comment (Magisus) [reply src/ruby/puppetserver-lib/puppet/server/logger.rb]: It appears to have previously been a string (see the last file in this diff).

Comment (MosesMendoza) [reply src/ruby/puppetserver-lib/puppet/server/logger.rb]: indeed

Comment (adrienthebo) [reply src/ruby/puppetserver-lib/puppet/server/logger.rb]: Puppetserver expects that this option will be set on the commandline and thus will be a string; if you pass a symbol Puppet explodes.

File: src/ruby/puppetserver-lib/puppet/server/logger.rb [hunk 3]
```diff
@@ -27,5 +27,25 @@ def handle(msg)
 class Puppet::Server::Logger
   def self.init_logging
     Puppet::Util::Log.newdestination(:logback)
+    Puppet[:log_level] = level_from_logback(get_logger)
```
Comment (MosesMendoza) [inline src/ruby/puppetserver-lib/puppet/server/logger.rb]: If I'm not mistaken, I think the trouble with this approach is that you can only set the log level once, when we start up. Future changes to the logback loglevel won't be reflected by the puppet loglevel.

I think we want something that queries the logback level when a log message is generated, so you can change it any time and have it be reflected across server & puppet.

File: src/ruby/puppetserver-lib/puppet/server/puppet_config.rb (hunk 1 above)
Comment (adrienthebo) [reply src/ruby/puppetserver-lib/puppet/server/puppet_config.rb]: This function is called to initialize Puppet within Puppetserver, but there was an existing function elsewhere that [enabled Puppetserver specific logging](https://github.com/puppetlabs/puppetserver/blob/master/src/ruby/puppetserver-lib/puppet/server/config.rb#L21), I just used that existing hook for setting the log level.

Comment (Magisus) [reply src/ruby/puppetserver-lib/puppet/server/puppet_config.rb]: I meant code to "control it via logback". Like, was there special filtering logic or something that we no longer need? (I don't really know anything about logback)

File: src/ruby/puppetserver-lib/puppet/server/logger.rb (hunk 3 above)
Comment (adrienthebo) [reply src/ruby/puppetserver-lib/puppet/server/logger.rb]: Under what conditions will the logback configuration be re-read? That file can't be read every time that a log message is generated, can it? As it stands if you reload the JRuby pool then the [Puppetserver master class](https://github.com/puppetlabs/puppetserver/blob/master/src/ruby/puppetserver-lib/puppet/server/master.rb#L28) should be regenerated and reloaded.

File: src/ruby/puppetserver-lib/puppet/server/puppet_config.rb (hunk 1 above)
Comment (adrienthebo) [reply src/ruby/puppetserver-lib/puppet/server/puppet_config.rb]: The :logback log destination works by grabbing the appropriately named logger and [relaying log messages](https://github.com/puppetlabs/puppetserver/blob/master/src/ruby/puppetserver-lib/puppet/server/logger.rb#L13-L23) to that instance. That logger instance itself should be determining if a message logged via `#info` should be dropped or actually logged.

Comment (Magisus) [reply src/ruby/puppetserver-lib/puppet/server/puppet_config.rb]: Cool thanks

File: src/ruby/puppetserver-lib/puppet/server/logger.rb (hunk 3 above)
Comment (bacongravy) [reply src/ruby/puppetserver-lib/puppet/server/logger.rb]: Does it need to automatically pick up the setting change, or could we document a manual step for users to perform after changing the setting?  Catching a signal (either SIGHUP, or SIGUSR1) is a common pattern I've seen for providing a way to tell a running process to re-read its configuration.  That said, signal handling is super tricky and if Puppet Server doesn't already catch those signals properly I wouldn't suggest adding code to do so at this point in the release cycle.

Comment (MosesMendoza) [reply src/ruby/puppetserver-lib/puppet/server/logger.rb]: Ah sure. I was under the impression you could dynamically update the log level in puppet server without having to do a server restart. In puppet you can just update the log level at any time and voila, new log level.

Comment (adrienthebo) [reply src/ruby/puppetserver-lib/puppet/server/logger.rb]: That's a fair point that Puppet can internally reconfigure logging but it's a dastardly thing to do; I don't think that we need to handle cases where some extension causes Puppet to change its log level at runtime.

I did some checking and it looks like logging is only configured when a [tk app is built](https://github.com/puppetlabs/trapperkeeper/blob/master/src/puppetlabs/trapperkeeper/core.clj#L44), so I'm not entirely sure that the trapperkeeper reload logic will re-read the logback.xml. However the `initialize_puppet_server` method will be invoked each time a [JRuby instance is created](https://github.com/puppetlabs/puppetserver/blob/master/src/clj/puppetlabs/services/jruby/jruby_puppet_core.clj#L139), so I think that the Ruby logic will be called more frequently than the logback.xml file will be read.

Comment (justinstoller) [reply src/ruby/puppetserver-lib/puppet/server/logger.rb]: TK will re-initialize logback during service reload. Logback has a "scan" feature that should check the file every (configurable) 60s and update its configuration if it's changed. We've had issues with logback's scan feature working reliably (see [TK-426](https://tickets.puppetlabs.com/browse/TK-426)).

Comment (adrienthebo) [reply src/ruby/puppetserver-lib/puppet/server/logger.rb]: welp

An alternative approach is to move the log level checking into the logger itself in `#handle` if it's essential to pick up these changes, but if TK will reinitialize logback during service reload then this implementation will pick up the log level changes at that point.

Comment (adrienthebo) [reply src/ruby/puppetserver-lib/puppet/server/logger.rb]: (I'm going to do some verification that reloading picks up the log level as expected.)

Comment (justinstoller) [reply src/ruby/puppetserver-lib/puppet/server/logger.rb]: You're right, finch. I was wrote that reply without seeing yours. I'm not seeing where tk re-initializes logging after a reload. (I guess I thought that [TK-424](https://tickets.puppetlabs.com/browse/TK-424) had been done).

File: src/ruby/puppetserver-lib/puppet/server/puppet_config.rb (hunk 1 above)
Comment (Sharpie) [reply src/ruby/puppetserver-lib/puppet/server/puppet_config.rb]: As long as adding:

```
<logger name="puppetserver" level="DEBUG"/>
```

To `/etc/puppetlabs/puppetserver/logback.xml` and restarting the puppetserver service still works, then we should be good from a user workflow perspective.

**Comment by MosesMendoza** [comment]
Seems reasonable. First we do `Puppet::Server::Config.initialize_puppet_server` which sets up the logback destination as its first task, then we call `Puppet::Server::PuppetConfig.initialize_puppet` in which we do all of the puppet setttings setup, and thus query logback for the log level and set puppet's log level correspondingly.

**Pull request merged**